File: 0_1.baekjoon/implementation/python/BOJ20165.py
from sys import stdin
readline = stdin.readline

N, M, R = map(int, readline().split())
board = [list(map(int, readline().split())) for _ in range(N)]
is_down = [[False for _ in range(M)] for _ in range(N)]
dir_map = {"E": 0, "W": 1, "S": 2, "N": 3}
D = ((0, 1), (0, -1), (1, 0), (-1, 0))

def solve():
    count = 0
    for _ in range(R):
        line = readline().split()
        row, col, _dir = int(line[0]) - 1, int(line[1]) - 1, dir_map[line[2]]
        
        if not is_down[row][col]:
            count += dfs(row, col, _dir, board[row][col]) + 1

        line = readline().split()
        row, col = int(line[0]) - 1, int(line[1]) - 1

        is_down[row][col] = False

    print(count)
    for i in range(N):
        for j in range(M):
            if is_down[i][j]:
                print("F", end=" ")
            else:
                print("S", end=" ")
        print()

def dfs(row, col, _dir, length):

    is_down[row][col] = True
    total = 0

    if length > 1:
        nr, nc = row + D[_dir][0], col + D[_dir][1]

        if 0 <= nr < N and 0 <= nc < M:
            if not is_down[nr][nc]:
                total += 1
                length = max(length - 1, board[nr][nc])
            else:
                length -= 1

            total += dfs(nr, nc, _dir, length)

    return total
        
solve() 

File: 0_1.baekjoon/implementation/java/src/BOJ20165.java
import java.io.*;
import java.util.*;

class BOJ20165 {
    static int N, M, R;
    static int[][] board;
    static boolean[][] isDown;
    static int[][] D = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());

        board = new int[N][M];
        isDown = new boolean[N][M];

        for (int i=0;i<N;i++) {
            st = new StringTokenizer(br.readLine());
            for (int j=0;j<M;j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        int result = 0;
        for (int i=0;i<R;i++) {
            st = new StringTokenizer(br.readLine());
            int row = Integer.parseInt(st.nextToken()) - 1;
            int col = Integer.parseInt(st.nextToken()) - 1;
            int dir = convertDirectionToInt(st.nextToken());
            
            if (!isDown[row][col])
                result += dfs(row, col, dir, board[row][col]) + 1;
            
            st = new StringTokenizer(br.readLine());
            row = Integer.parseInt(st.nextToken()) - 1;
            col = Integer.parseInt(st.nextToken()) - 1;

            if (isDown[row][col]) {
                isDown[row][col] = false;
            }
        }

        printResult(result);

    }
    private static void printResult(int result) {
        System.out.println(result);
        for (boolean[] down : isDown) {
            for (boolean d : down) {
                if (d) {
                    System.out.print("F");
                } else {
                    System.out.print("S");
                }
                System.out.print(" ");
            }
            System.out.println();
        }
    }
    private static int dfs(int row, int col, int dir, int height) {
        
        isDown[row][col] = true;
        int total = 0;
        if (height > 1) {
            int nr = row + D[dir][0], nc = col + D[dir][1];

            if (0 <= nr && nr < N && 0 <= nc && nc < M) {
                if (!isDown[nr][nc]) {
                    height = Math.max(height - 1, board[nr][nc]);
                    total++;
                } else {
                    height--;
                }

                total += dfs(nr, nc, dir, height);
            }
        }
        return total;
        
    }
    private static int convertDirectionToInt(String direction) {
        if (direction.equals("E"))
            return 0;

        if (direction.equals("W"))
            return 1;

        if (direction.equals("S"))
            return 2;

        return 3;
    }
}